src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -235,19 +277,24 @@ private void InitializeConstructorParameters(Dictionary<string, JsonPropertyInfo
 
             foreach (ParameterInfo parameterInfo in parameters)
             {
-                PropertyInfo? firstMatch = null;
+                MemberInfo? firstMatch = null;
                 bool isBound = false;
 
                 foreach (JsonPropertyInfo jsonPropertyInfo in PropertyCacheArray!)
                 {
                     // This is not null because it is an actual
                     // property on a type, not a "policy property".
-                    PropertyInfo propertyInfo = jsonPropertyInfo.PropertyInfo!;
+                    MemberInfo memberInfo = jsonPropertyInfo.MemberInfo!;
+                    Debug.Assert(memberInfo is PropertyInfo || memberInfo is FieldInfo);
+
+                    Type memberType = memberInfo is PropertyInfo propertyInfo
+                        ? propertyInfo.PropertyType
+                        : Unsafe.As<FieldInfo>(memberInfo).FieldType;
```

Comment:
Is there any measurable perf benefit to using `Unsafe.As` here over a regular cast?

Reply:
One less cast. This trick is used by `ValueTask` for example, spans have it too.

Personally I would like to avoid these repeatable lines of code, but there is no other way to get the true property type because placeholders use `byte`. I told you about my concerns on that before.

A good way to avoid this is to have `JsonPlaceholderPropertyInfo` which won't be generic, will implement all required stuff to represent a property as it is, but which will throw on any attempt to read or write a value. In that case it will be fine to use `JsonPropertyInfo.DeclaredPropertyType` or `RuntimePropertyType`.